### PR TITLE
feat(tracing): add support for grpc.aio

### PR DESCRIPTION
**Pull request by ysk24ok**
Took over #2958 and made tracer for `grpc.aio` work.

Tests are executed w/ Python3.7, 3.8, 3.9 and 3.10.
Since `grpc.aio` graduated from experimentation in [1.32.0](https://github.com/grpc/grpc/releases/tag/v1.32.0)
- As for Python 3.7 and 3.8, minimum version is 1.32.0
- As for Python 3.9, minimum version is 1.33.1 because wheels are not provided in 1.32.0
- As for Python3.10, minimum version is 1.42.0 where deprecated `loop` variable was resolved

## Checklist
- [ ] Added to the correct milestone.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

Not sure how to add correct milestone and add library documentation. Would be great if you tell me. Thank you.

### CODE REVIEW

**Comment by ysk24ok** [comment]
In `grpc_aio` job, the error say "The specified paths did not match any files in /home/circleci/project/coverage".
I guess this is because tests are not executed in the previous step. I have no idea why.

File: ddtrace/contrib/grpc_aio/aio_client_interceptor.py
Comment (isac322) [inline ddtrace/contrib/grpc_aio/aio_client_interceptor.py:30]:
```suggestion
    # type: (Pin, str, int) -> Tuple[aio.ClientInterceptor, ...]
```

I think it should be changed

File: ddtrace/contrib/grpc_aio/aio_client_interceptor.py
Comment (isac322) [inline ddtrace/contrib/grpc_aio/aio_client_interceptor.py:40]:
```suggestion
    # type: (Span, grpc.StatusCode, str) -> Callable[[aio.Call], None]
```

**Comment by isac322** [comment]
And I think it should be consider non coroutine handler. grpcio accept `migration_thread_pool` to handle non-coroutine handler in aio server. (https://grpc.github.io/grpc/python/grpc_asyncio.html#create-server)
I got below error on aio server have non-coroutine handler.

```
Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 646, in grpc._cython.cygrpc._handle_exceptions
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 745, in _handle_rpc
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 511, in _handle_unary_unary_rpc
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 368, in _finish_handler_with_unary_response
  File "/src/statssrv/grpc_aio/aio_server_interceptor.py", line 115, in unary_unary
    return await _wrap_unary_response(self.__wrapped__.unary_unary, request, context, span)
  File "/src/statssrv/grpc_aio/aio_server_interceptor.py", line 67, in _wrap_unary_response
    return await behavior(request_or_iterator, servicer_context)
  File "/src/statssrv/server.py", line 264, in GetCountryAverageInventory
    res = views.get_country_average_inventory(**req)
  File "/src/stats/views.py", line 57, in wrapped
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/cacheops/simple.py", line 57, in wrapper
    result = func(*args, **kwargs)
  File "/src/stats/views.py", line 271, in get_country_average_inventory
    res = get_average_inventory_by_where(
  File "/src/stats/views.py", line 232, in get_average_inventory_by_where
    with connection.cursor() as cursor:
  File "/usr/local/lib/python3.8/site-packages/ddtrace/contrib/trace_utils.py", line 162, in wrapper
    return func(mod, pin, wrapped, instance, args, kwargs)
  File "/usr/local/lib/python3.8/site-packages/ddtrace/contrib/django/patch.py", line 85, in cursor
    cursor = func(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/django/utils/asyncio.py", line 24, in inner
    raise SynchronousOnlyOperation(message)
django.core.exceptions.SynchronousOnlyOperation: You cannot call this from an async context - use a thread or sync_to_async.
```

**Comment by ysk24ok** [comment]
@brettlangdon Hello.

> And I think it should be consider non coroutine handler. grpcio accept migration_thread_pool to handle non-coroutine handler in aio server.

The problem described in https://github.com/DataDog/dd-trace-py/pull/3240#issuecomment-1036312734 is now resolved by https://github.com/ysk24ok/dd-trace-py/pull/1 and https://github.com/ysk24ok/dd-trace-py/pull/2, so it would be great if you could allocate some time to review this.
`grpc_aio` test succeeds locally, but it fails in CI.

File: ddtrace/contrib/grpc_aio/__init__.py
Comment (brettlangdon) [inline ddtrace/contrib/grpc_aio/__init__.py:6]:
```suggestion
~~~~~~~~

To manually enable the integration::
```

File: .circleci/config.yml
Comment (brettlangdon) [inline .circleci/config.yml:624]: you probably don't need this, since the `grpc_aio` tests will get picked up by `pattern: "grpc"` above.

or we can do:

```
pattern: "^grpc$"

pattern: "^grpc_aio$"
```

to separate them

**Review by brettlangdon — COMMENTED** [review]
it isn't obvious to me what is going on, but it looks like CircleCI isn't picking up any of the riot venvs to run.

https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/15142/workflows/0189b7c4-3110-4a61-b80e-eae0757e8560/jobs/1049598/parallel-runs/2/steps/2-106

```
#!/bin/bash -eo pipefail
mv .riot .ddriot
./scripts/tested_pythons | circleci tests split | xargs -I PY ./scripts/ddtest riot -v run --python=PY --exitfirst --pass-env -s 'grpc_aio'

WARNING: The TMPDIR variable is not set. Defaulting to a blank string.
Pulling testrunner (datadog/dd-trace-py:buster)...
buster: Pulling from datadog/dd-trace-py

5158dd02: Pulling fs layer
ba1b6771: Pulling fs layer
5d7dc0eb: Pulling fs layer
058814c2: Pulling fs layer
83d22b29: Pulling fs layer
Digest: sha256:e35afc78dd38c63673660bf04a1e30a03b4e55377029a71cd12cf359e0f9b6b1  28.3MB/28.3MBBB
Status: Downloaded newer image for datadog/dd-trace-py:buster
Creating project_testrunner_run ... 
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
[20:54:14] INFO     Generating virtual environments for interpreters riot.py:831

-------------------summary-------------------
0 passed with 0 warnings, 0 failed
```

File: .circleci/config.yml
Comment (brettlangdon) [reply .circleci/config.yml]: 🤔 that didn't seem to even fix it. one of us will have to try this locally and see if we can figure out why it isn't finding/running any of the `grpc_aio` test cases.

Comment (ysk24ok) [reply .circleci/config.yml]: Now I got why.

`scripts/tested_pythons` says there are 7 python versions (i.e. 2.7, 3.5-3.10) and `circleci tests split` splits them between workers so when parallelism is 6, 1 worker processes almost 1 python version. Since not all of the test results are shown in the CIrcleCI UI and `grpc_aio` job does not have tests for 2.7 and 3.5, CI ends up 0 executed tests.

After I reduced the parallelism to 2 in 87fa759, [`grpc_aio` job](https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/15178/workflows/79ab5b56-e520-4028-af89-d28824f26bf0/jobs/1051739) worked fine. NOTE THAT it only contains test results for 3.7 and 3.9 because tests are splitted across workers and not all of them are shown in UI.

Is this a known issue? If not, I'll create an issue.

Comment (brettlangdon) [reply .circleci/config.yml]: I see 3.7, 3.8, 3.9, and 3.10 in the CircleCI UI.

You have to scroll through the results from both workers to catch all the pytest runs, we don't have an executive summary at the end of a `riot` run.

**Comment by brettlangdon** [comment]
@ysk24ok can you add a release note to this PR?

```
pip install -U reno
reno new add-grpc-aio-support
```

This would be a "features:" note.

(we still need someone to do a full review of the PR, but glad we got CI working, and can get the other checks passing)

File: .circleci/config.yml
Comment (ysk24ok) [reply .circleci/config.yml]: Screen Shot 2022-03-09 at 6 52 06
Oh, I didn't know that we could see the results of workers.

Anyway, the reason why CI failed was there were some workers to which valid Python version was not allocated.

**Comment by ysk24ok** [comment]
@brettlangdon Added a release note in [f965d67](https://github.com/DataDog/dd-trace-py/pull/3240/commits/f965d6731f77f15ba454d6858fe080f232d2a573).

**Review by majorgreys — COMMENTED** [review]
@ysk24ok I was wondering what your reasons were adding this as a separate integration from `grpc`. Given the patching here, we would, by default, patch both the sync and async grpc client and server.

**Comment by ysk24ok** [comment]
@majorgreys I just took over #2958 @tu-artem created, maybe we should ask him.

If maintainers think it's better to integrate `grpc` and `grpc_aio`, I'll work on it.

**Comment by tu-artem** [comment]
@majorgreys I think the only reason was particular version of `grpc` when `aio` became available, so it was just easier to have it separately

**Comment by ysk24ok** [comment]
@brettlangdon @majorgreys 
Based on @tu-artem's comment, would it be better to integrate `grpc` and `grpc_aio`?
I would say we could do by patching `grpc.aio` only when `grpc >= 1.32`.

**Comment by ysk24ok** [comment]
@brettlangdon @majorgreys 
Hello, I've integrated `grpc_aio` into `grpc` and now `patch` method patches both modules.
Note that as for tests `grpc` and `grpc_aio` remain separated because we don't need run `grpc_aio` with Python2 and `grpcio<1.32.0`.
Could you take a look? Some CI jobs failed but I believe this PR has any bearing on them.

**Comment by ysk24ok** [comment]
@brettlangdon @majorgreys 
Sorry to ping multiple times, but could you tell me when you're going to review this PR?
We're planning to incorporate this feature into the library developed in our company and I'd like to decide whether I have to wait for your review and the new release, or directly use this branch.

**Comment by majorgreys** [comment]
@ysk24ok We are sorry we haven not gotten to this PR sooner. We were focused on getting v1.0 out the past few weeks. I will review this PR today and get back to you by end of day!

File: releasenotes/notes/add-grpc-aio-support-c59efa5557aeea8f.yaml
Comment (majorgreys) [inline releasenotes/notes/add-grpc-aio-support-c59efa5557aeea8f.yaml:4]:
```suggestion
    Add support for ``grpc.aio``.
```

File: riotfile.py
Comment (majorgreys) [inline riotfile.py:1323]:
```suggestion
                        # 3.9 wheels are not provided in 1.32.0
                        "grpcio": ["~=1.33.0", latest],
```

File: riotfile.py
Comment (majorgreys) [inline riotfile.py:1329]: I did not follow the point made in the PR description for why this is necessary. Can you add a comment and a link to any relevant source/issue that explains why we have this condition?

File: riotfile.py
Comment (majorgreys) [inline riotfile.py:1311]: As you can see from the existing tests, we include minor versions in the test matrix. This minor version is from Sep 2020. Were there any other minor versions where we want to this code with `grpcio.aio`? Or do you think we should be testing some or all of the intervening minor versions?

**Review by majorgreys — COMMENTED** [review]
Generally looks good, just a few clarifications that could be useful as comments in the code. 

Also, the `grpc_aio` job needs to be added to the `.circleci/config.yml`.

File: riotfile.py
Comment (ysk24ok) [reply riotfile.py:1324]: Added a comment in 0890e71.

File: riotfile.py
Comment (ysk24ok) [reply riotfile.py:1332]: Added a comment in 0890e71.

File: ddtrace/contrib/grpc/aio_server_interceptor.py
Comment (ysk24ok) [reply ddtrace/contrib/grpc/aio_server_interceptor.py]: Added a comment and did a little refactoring in 86f24f3.

File: ddtrace/contrib/grpc/aio_client_interceptor.py
Comment (ysk24ok) [reply ddtrace/contrib/grpc/aio_client_interceptor.py:192]: Added comments for `_wrap_stream_response` and `_wrap_unary_response` in a10a038.
Using `functools.partial` is not a big concern, it's just used to reduce the number of arguments to be passed to `_wrap_unary_response`.

File: ddtrace/contrib/grpc/aio_client_interceptor.py
Comment (ysk24ok) [reply ddtrace/contrib/grpc/aio_client_interceptor.py:147]: As far as I tested locally, directly running post-processing here instead of setting the callback worked fine. I used `add_done_callback` because it was also used in existing `ddtrace/contrib/grpc`.
Should I directly call post-processing here?

File: riotfile.py
Comment (ysk24ok) [reply riotfile.py:1311]: Might be better to add minor versions of grpcio which has any update in `grpc.aio`.

**Comment by ysk24ok** [comment]
@majorgreys Thank you for your review! I've made some updates, so it would be great if you could take a look.

> Also, the grpc_aio job needs to be added to the .circleci/config.yml. You should be able to copy the grpc job:

The regexp is `grpc`, not `grpc$`, so the CI job also executes tests for `grpc.aio`. See [this job](https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/16330/workflows/f47c2b47-8c1c-4dd7-b2c2-477df5229d94/jobs/1120490) and you'll find out the tests are also executed.

File: ddtrace/contrib/grpc/aio_client_interceptor.py
Comment (majorgreys) [reply ddtrace/contrib/grpc/aio_client_interceptor.py:147]: I would use the callback but add a comment about why the implementation has the callback registered after the iterator is finished.

File: riotfile.py
Comment (majorgreys) [reply riotfile.py:1311]: Right, sorry I realize what I wrote was not clear. We can tackle adding more versions to the test matrix as a follow-up.

File: ddtrace/contrib/grpc/constants.py
Comment (brettlangdon) [inline ddtrace/contrib/grpc/constants.py:10]:
```suggestion
GRPC_AIO_PIN_MODULE_SERVER = None
GRPC_AIO_PIN_MODULE_CLIENT = None
try:
    from grpc import aio

    GRPC_AIO_PIN_MODULE_SERVER = aio.Server
    GRPC_AIO_PIN_MODULE_CLIENT = aio.Channel
except ImportError:
    pass
```

? are we going to have any issues if these are not defined?

We do use as `getattr(constants.GRPC_AIO_PIN_MODULE_SERVER, ... )`

Maybe these are better off not as constants?

e.g.

```python
def _patch_aio_client():
    if getattr(grpc.aio.Channel, "__datadog_patch", False):
        return
    setattr(grpc.aio.Channel, "__datadog_patch", True)

    Pin().onto(grpc.aio.Channel)
```

Since we only call `_patch_aio_client()` if `HAS_GRPC_AIO is True`.

what do you think?

File: ddtrace/contrib/grpc/aio_client_interceptor.py
Comment (ysk24ok) [reply ddtrace/contrib/grpc/aio_client_interceptor.py:149]: Made a comment in 14dc78a.

File: ddtrace/contrib/grpc/aio_client_interceptor.py
Comment (ysk24ok) [reply ddtrace/contrib/grpc/aio_client_interceptor.py:175]: Made a comment in 14dc78a.

File: ddtrace/contrib/grpc/constants.py
Comment (ysk24ok) [reply ddtrace/contrib/grpc/constants.py]: Updated in c716a6d. Would be great if you could have a look.
Instead of using raw `grpc.aio.Channel` and `grpc.aio.Server`, I decided to move the constants to `patch.py` as they are used in the test as well. This led to remove duplicated try-except to check if `grpc.aio` is available and make comments why they are not defined in `constants.py`.

File: ddtrace/contrib/grpc/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/grpc/patch.py:235]:
```suggestion
    pin = Pin.get_from(GRPC_AIO_PIN_MODULE_SERVER)
```

did you also want to use the constant here?

Comment (ysk24ok) [reply ddtrace/contrib/grpc/patch.py]: Thank you, at first I tried to replace the constants with `grpc.aio.XXX` and I guess this was left behind...

File: ddtrace/contrib/grpc/aio_server_interceptor.py
Comment (brettlangdon) [inline ddtrace/contrib/grpc/aio_server_interceptor.py:37]: This is more a comment for @majorgreys 

I know this code only works on newer Python, but should we use a type comment here to be consistent?

I am not sure I have a strong opinion, just calling it out if we want to make a decision here or not.

Comment (ysk24ok) [reply ddtrace/contrib/grpc/aio_server_interceptor.py:37]: Maybe I shouldn't have added the type hint because in other codes type comments are used so that they can run with old Python.
Since the name of the variable `_INT2CODE` almost represents its type, no type hint would be fine.

Comment (ysk24ok) [reply ddtrace/contrib/grpc/aio_server_interceptor.py]: Deleted the type hint in 1d081c6.

**Review by majorgreys — APPROVED** [review]
🙇🏽

File: ddtrace/contrib/grpc/aio_server_interceptor.py
Comment (brettlangdon) [reply ddtrace/contrib/grpc/aio_server_interceptor.py]: I actually think it was probably fine, since this is a Python 3 specific file we probably have typing available natively, but it is also ok that you removed it!

**Review by brettlangdon — APPROVED** [review]

**Comment by ysk24ok** [comment]
@brettlangdon @majorgreys 
Thank you for your approval!
Can I ask when v1.2.0 is planned to be released?

**Comment by brettlangdon** [comment]
@Mergifyio refresh

**Pull request merged**
